### PR TITLE
TRPL: deref coercions

### PR DESCRIPTION
**Pull request by steveklabnik**
r? @alexcrichton

### CODE REVIEW

File: src/doc/trpl/deref-coercions.md [hunk 1]
```diff
@@ -1,3 +1,73 @@
...
+    assert_eq!('a', *x);
+}
+```
+
+[deref]: ../std/ops/trait.Deref.html
+
+This is useful for writing custom pointer types. However, there’s a language
+feature related to `Deref`: ‘deref coercions’. Here’s the rule: If you have a
+type `U`, and it implements `Deref<Target=T>`, values of `&U` will
+automatically convert to a `&T`. Here’s an example:
```
Comment (alexcrichton) [inline src/doc/trpl/deref-coercions.md]: s/convert/coerce/ 

I think the distinction may matter in some places? I'm not super sure though

**Comment by alexcrichton** [comment]
Nice! Could this also mention somewhere that `Vec<T>` will coerce to `[T]`? I think that's the only other major coercion that may need to be mentioned.

**Comment by alexcrichton** [comment]
r=me with those nits otherwise though!

**Comment by Manishearth** [comment]
Can we also mention (or link to) autoderef here somewhere? I'd like all of the deref magic to be mentioned in one place.

**Comment by steveklabnik** [comment]
issues fixed, and @Manishearth 's section added :smile:

File: src/doc/trpl/deref-coercions.md [hunk 2]
```diff
@@ -1,3 +1,117 @@
...
+f.foo();
+```
+
+Even though `f` isn’t a reference, and `foo` takes `&self`, this works.
+That’s because these things are the same:
+
+```rust,ignore
+f.foo();
+(*f).foo();
+(******f).foo();
```
Comment (alexcrichton) [inline src/doc/trpl/deref-coercions.md]: Did this mean to use `&` instead of `*`?

Comment (steveklabnik) [inline src/doc/trpl/deref-coercions.md]: Yes. eff.

Comment (steveklabnik) [inline src/doc/trpl/deref-coercions.md]: well, i guess it actually meant to use `&` and `*` alternating, since that's the behavior, right?

Comment (alexcrichton) [inline src/doc/trpl/deref-coercions.md]: I think this means to say that a value of type `&&&&&&&&&&&&&&&&Foo` can still have methods defined on `Foo` called as the compiler will insert as many `*` operations as necessary (basically).

Comment (steveklabnik) [inline src/doc/trpl/deref-coercions.md]: I'm thinking of http://stackoverflow.com/questions/28519997/what-are-rusts-exact-auto-dereferencing-rules/28552082#28552082

**Comment by steveklabnik** [comment]
adding to https://github.com/rust-lang/rust/pull/24786

File: src/doc/trpl/deref-coercions.md [hunk 3]
```diff
@@ -1,3 +1,119 @@
...
+
+foo(&owned);
+```
+
+Vectors can `Deref` to a slice.
+
+## Deref and method calls
+
+`Deref` will also kick in when calling a method. In other words, these are
+the same two things in Rust:
```
Comment (frewsxcv) [inline src/doc/trpl/deref-coercions.md]: Which 'two things' is this referring to?

**Pull request merged**